js.go

```diff
@@ -1214,6 +1364,262 @@ type StreamState struct {
+		return err
+	}
+	var resp JSApiStreamDeleteResponse
+	if err := json.Unmarshal(r.Data, &resp); err != nil {
+		return err
+	}
+	if resp.Error != nil {
+		return errors.New(resp.Error.Description)
+	}
+	return nil
+}
+
+type StreamSnapshotConfig struct {
+	// Subject to deliver the chunks to for the snapshot.
+	DeliverSubject string `json:"deliver_subject"`
+	// Do not include consumers in the snapshot.
+	NoConsumers bool `json:"no_consumers,omitempty"`
+	// Optional chunk size preference.
+	// Best to just let server select.
+	ChunkSize int `json:"chunk_size,omitempty"`
+	// Check all message's checksums prior to snapshot.
+	CheckMsgs bool `json:"jsck,omitempty"`
+}
+
+// JSApiStreamSnapshotResponse is the direct response to the snapshot request.
+type JSApiStreamSnapshotResponse struct {
+	APIResponse
+	// Estimate of number of blocks for the messages.
+	NumBlks int `json:"num_blks"`
+	// Block size limit as specified by the stream.
+	BlkSize int `json:"blk_size"`
+}
+
+func (js *js) SnapshotStream(name string, cfg *StreamSnapshotConfig) (io.Reader, error) {
+	req, err := json.Marshal(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	ssSubj := js.apiSubj(fmt.Sprintf(JSApiStreamSnapshotT, name))
+	r, err := js.nc.Request(ssSubj, req, js.wait)
+	if err != nil {
+		return nil, err
+	}
+	var resp JSApiStreamSnapshotResponse
+	if err := json.Unmarshal(r.Data, &resp); err != nil {
+		return nil, err
+	}
+	if resp.Error != nil {
+		return nil, errors.New(resp.Error.Description)
+	}
+
+	// Setup to process snapshot chunks.
+	var snapshot []byte
+	done := make(chan bool)
+	sub, err := js.nc.Subscribe(cfg.DeliverSubject, func(m *Msg) {
+		// EOF
+		if len(m.Data) == 0 {
+			done <- true
+			return
+		}
+		// Could be writing to a file here too.
+		snapshot = append(snapshot, m.Data...)
+		// Flow ack
+		m.Respond(nil)
+	})
+	if err != nil {
+		return nil, err
+	}
+	defer sub.Unsubscribe()
+	// Wait to receive the snapshot.
+	select {
+	case <-done:
+	case <-time.After(5 * time.Second):
+		return nil, errors.New("nats: snapshot timeout exceeded")
+	}
+
+	return bytes.NewReader(snapshot), nil
+}
+
+// JSApiStreamRestoreResponse is the direct response to the restore request.
+type JSApiStreamRestoreResponse struct {
+	APIResponse
+	// Subject to deliver the chunks to for the snapshot restore.
+	DeliverSubject string `json:"deliver_subject"`
+}
+
+func (js *js) RestoreStream(name string, snapshot io.Reader) error {
+	rsSubj := js.apiSubj(fmt.Sprintf(JSApiStreamRestoreT, name))
+	r, err := js.nc.Request(rsSubj, nil, js.wait)
+	if err != nil {
+		return err
+	}
+	var resp JSApiStreamRestoreResponse
+	if err := json.Unmarshal(r.Data, &resp); err != nil {
+		return err
+	}
+	if resp.Error != nil {
+		return errors.New(resp.Error.Description)
+	}
+
+	// Can be any size message.
+	var chunk [512]byte
+	for r := snapshot; ; {
+		n, err := r.Read(chunk[:])
+		if err != nil {
+			break
+		}
+		js.nc.Request(resp.DeliverSubject, chunk[:n], time.Second)
```

Comment:
Same, should this be configurable? Or use the `js.wait`? I know @wallyqs added a wait to make the wait being generic and used in various places.